File: rust/8.string-to-integer-atoi.rs
/*
 * @lc app=leetcode id=8 lang=rust
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
impl Solution {
    pub fn my_atoi(s: String) -> i32 {
        let s = s.trim_start();
        let mut i = 0;
        let mut sign = 1;
        let mut chars = s.chars();
        if let Some(c) = chars.next() {
            match c {
                '-' => sign = -1,
                '+' => sign = 1,
                _ => {
                    if c.is_ascii_digit() {
                        i += c.to_digit(10).unwrap() as i32;
                    } else {
                        return i;
                    }
                }
            }
        }
        while let Some(c) = chars.next() {
            if c.is_ascii_digit() {
                if let Some(m) = i.checked_mul(10) {
                    if let Some(a) = m.checked_add(c.to_digit(10).unwrap() as i32) {
                        i = a;
                        continue;
                    }
                }
            } else {
                break;
            }
            if sign == 1 {
                return i32::MAX;
            }
            return i32::MIN;
        }
        return i * sign;
    }
}
// @lc code=end

File: javascript/8.string-to-integer-atoi.js
/*
 * @lc app=leetcode id=8 lang=javascript
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function (s) {
  const MIN_NUM = -(2 ** 31),
    MAX_NUM = 2 ** 31 - 1;
  let r = "0";
  let i = 0;
  while (i < s.length && s[i] === " ") i++;
  if (s[i] === "-") {
    r = "-" + r;
    i++;
  } else if (s[i] === "+") i++;
  while (i < s.length && s[i] !== " " && !isNaN(s[i])) {
    r += s[i++];
    if (r > MAX_NUM) return MAX_NUM;
    else if (r < MIN_NUM) return MIN_NUM;
  }
  return r;
};
// @lc code=end
